app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,303 @@
+            }
+            directionKeysPressed.isEmpty() -> {
+                resetCursorSpeed()
+                return false
+            }
+            else -> {
+                val currTime = System.currentTimeMillis()
+                // If the cursor pos has not recently been updated, reset to current time.
+                // Otherwise, the cursor would shoot across the page the first time it was
+                // touched
+                if (lastUpdatedAtMS == NOT_RECENTLY_UPDATED) lastUpdatedAtMS = currTime - MS_PER_FRAME
+                lastVelocity = internalMutatePositionAndReturnVelocity(
+                        oldPos,
+                        lastUpdatedAtMS,
+                        currTime,
+                        lastVelocity,
+                        directionKeysPressed
+                )
+                lastUpdatedAtMS = currTime
+                return true
+            }
+        }
+    }
+
+    /**
+     * Mutates [oldPos] to its new position. This position is calculated based on the time passed
+     * since the last update, in order to avoid miscalculations when dropping frames.
+     *
+     * Calculates necessary scrolling, if any, and sends out a request to handle it.
+     *
+     * @returns the new cursor velocity
+     */
+    private fun internalMutatePositionAndReturnVelocity(
+        oldPos: PointF,
+        oldTimeMS: Long,
+        newTimeMS: Long,
+        oldVelocity: Float,
+        directionsPressed: Set<Direction>
+    ): Float {
+        // directionsPressed empty case is handled in `mutatePosition`
+        require(directionsPressed.isNotEmpty())
+        val screenBounds = screenBounds ?: return 0f
+
+        val timePassed = newTimeMS - oldTimeMS
+        val accelerateBy = ACCELERATION_PER_MS * timePassed
+        val velocity = (oldVelocity + accelerateBy).coerceIn(0f, MAX_VELOCITY)
+
+        fun updatePosition() {
+            var verticalVelocity = 0f
+            if (directionKeysPressed.contains(Direction.UP)) verticalVelocity -= velocity
+            if (directionKeysPressed.contains(Direction.DOWN)) verticalVelocity += velocity
+            var horizontalVelocity = 0f
+            if (directionKeysPressed.contains(Direction.LEFT)) horizontalVelocity -= velocity
+            if (directionKeysPressed.contains(Direction.RIGHT)) horizontalVelocity += velocity
+            oldPos.x += horizontalVelocity
+            oldPos.y += verticalVelocity
+            oldPos.x = oldPos.x.coerceIn(0f, screenBounds.x)
+            oldPos.y = oldPos.y.coerceIn(0f, screenBounds.y)
+        }
+
+        fun calculateAndSendScrollEvent() {
+            // This should not live inside internalMutatePositionAndReturnVelocity.
+            // Move it if you can find a better solution
+            val approxFramesPassed = (timePassed / 16).toInt()
+            val shouldScroll = getScrollDistance(velocity, oldPos, approxFramesPassed)
+            shouldScroll?.let { scrollBus.scrollRequests.onNext(it) }
```

Comment:
I think this would be acceptable, and fit our app's architecture, if the `scrollBus` Observable lived inside `CursorModel`. Do you agree?

Reply:
That's definitely an improvement.  I still don't like it, because it's a very similar conceptual space to what `CursorEventRepo` handles, but I spent some time trying to unify them and wasn't able to find any easy solution to that.

So I'll take your advice, good call.

Reply:
tbh, I'm still confused on the distinction between `CursorEventRepo` and `CursorModel`: maybe it's this same dissonance that makes it challenging to combine the `ScrollRepo` with `CursorEventRepo`?

Given what we learned from the original implementation (having an arbitrary architecture that is even slightly unclear is hard to maintain because eventually code will be added in the wrong place then it will be even more unclear), I wonder if there's value in merging CursorEventRepo directly into CursorModel: it'd make the responsibilities much clearer between the two remaining classes (Model & View). Once we better understand the problem space and the `CursorModel` grows too large, we can pull whatever functionality makes sense into a separate class.

I'm not sure this is a good idea, just thinking aloud...

Reply:
That's an interesting observation, and I think it might be a good suggestion.  I think I need to mull it over before making that change, though.

Reply:
It'd also address the circular dependency between CursorEventRepo and CursorModel (if we were to move EventRepo inside), though there may be other solutions for that.